### PR TITLE
feat(mobile): show local assets

### PR DESCRIPTION
**Pull request by fyfrey**
This PR is a first step towards using Immich offline: Displaying on-device assets that have not been backup up yet 

As this is already a huge change set, I do not plan to add support for unified assets (link local and remote assets) in this PR. This should go into a follow-up PR.

Changes:
- [x] introduce new class `Asset` as composition of `AssetResponseDto` and `AssetEntity`
- [x] introduce new class `ImmichImage` to render `Asset` automatically using either local or remote data (using `CachedNetworkImage`)
- [x] removed quite some code duplication from using `CachedNetworkImage` in various places with similar setttings 
- [x] Shows all remote assets (still loaded from server)
- [x] Shows all local assets that are newer than the last successful backup minus those that are already backed up
- [x] find out what to do with local images that already exist on the server (same hash, but different `deviceAssetId` or `deviceId`, occurs when installing Immich on another device and moving assets to the other device) -> nothing to be done unless we want to do client-side hashing as well, resolves once backup is performed (because only local assets newer than last backup time are shown)  
- [x] filter out duplicate assets (that are both local and remote, take only remote for now)
- [x] only allow remote images to be added to albums etc.
- [x] support local videos (proper playback and loading indicator)
- [x] `getAllAsset` waits until BackgroundService is done before opening Hive `HiveBackupAlbums` box
- [x] change storage indicator icons in asset grid: cloud (only remote asset), cloud_done (asset uploaded from this device), cloud_off (local asset that has not yet been uploaded) 
- [x] fix deletion of assets and drastically improve delete speed on large asset collections
- [x] fix `onNewAssetUploaded`: The local asset (if found) is now deleted (only in the app state, not physically!) and the remote asset is added to the app state

### CODE REVIEW

File: mobile/lib/modules/home/services/asset.service.dart [hunk 1]
```diff
@@ -1,34 +1,73 @@
...
 
-  Future<List<AssetResponseDto>?> getAllAsset() async {
+  Future<List<Asset>> getAllAsset() async {
+    final List<Asset> assets = [];
     try {
-      return await _apiService.assetApi.getAllAssets();
+      final Future<List<AssetResponseDto>?> remoteTask =
+          _apiService.assetApi.getAllAssets();
+
+      /* FIXME this is bad:
```
Comment (zoodyy) [inline mobile/lib/modules/home/services/asset.service.dart:35]: how do we want to tackle this?

1. ignore the potential problem (currently implemented): doing this discards changes made by the background service to `HiveBackupAlbums` because the foreground app already has the data loaded
2. somehow wait until BackgroundBackup has finished (non-trivial to do): If the local cache is empty or invalid, we don't show any assets until the BackgroundBackup is done.

Comment (alextran1502) [reply mobile/lib/modules/home/services/asset.service.dart:35]: If we wait for the BackgroundBackup to finish, it means the foreground UI will be locked, correct?

Comment (zoodyy) [reply mobile/lib/modules/home/services/asset.service.dart:35]: I tested this with a `await Future.delayed(const Duration(seconds: 30));` and you can actually still use the app.
One issue is... depending on what the user does `getAllAsset` is called multiple times and all these calls are blocked and continue after waiting

Comment (alextran1502) [reply mobile/lib/modules/home/services/asset.service.dart:35]: How does it cause problems for this PR? I think I miss somethings here to understand the problem entirely.

Comment (zoodyy) [reply mobile/lib/modules/home/services/asset.service.dart]: found a way so that multiple places in the code can wait on the background service.

for `getAllAssets`: if the cached state is invalid, it will at first not show local assets when background service is actually running - only remote assets

**Comment by zoodyy** [comment]
> It works great on the emulator and my little test instance! However, when I install the app on my phone that uses my prod-instance with around 20k assets, it shows local assets and the same asset with the synced-icon even after sync is complete. Maybe some kind of race-condition?

Thanks for testing it @matthinc !

I also noticed that when performing backup in dev mode and checking the release build later on. Turns out: release and debug builds have different deviceIDs! Did you try a debug or release build? If it was debug.. please try again with a release build.

File: mobile/lib/shared/providers/asset.provider.dart [hunk 2]
```diff
@@ -52,53 +66,67 @@ class AssetNotifier extends StateNotifier<List<AssetResponseDto>> {
   }
 
   onNewAssetUploaded(AssetResponseDto newAsset) {
-    state = [...state, newAsset];
+    state = [...state, Asset.remote(newAsset)];
```
Comment (matthinc) [inline mobile/lib/shared/providers/asset.provider.dart:69]: This is called by the ws connection. Maybe the duplication occurs because we don't compare the id of the new asset agains the ids of our local assets here?

**Comment by alextran1502** [comment]
Hi @zoodyy, thank you for another high-quality and high-value PR.

## Observation before looking in the code

1. The initial loading speed of the app is fantastic. I think this is from the fewer assets that need to be loaded if the assets exist on the device, correct?
2. I saw we have three different icons to indicate the state of an asset. I think you planned to use two icons, "cloud with the cross mark - indicates the asset available locally and not on remote" and "cloud with the check mark - indicates the asset available remotely." Regarding the other icon - "just a cloud" I don't know what state the asset is in. I observed it in my testing instance, and the asset is confirmed to be available locally and remotely. Please see the screenshot below

IMG_8972

3. Now that we need to use the icon to show the state of the photos. I think the icon will need to be presented at all times and not optional anymore - @matthinc, thoughts?

**Comment by matthinc** [comment]
@alextran1502 Yes, I agree. The icon plays a very crucial role now.

File: mobile/lib/shared/providers/asset.provider.dart (hunk 2 above)
Comment (zoodyy) [reply mobile/lib/shared/providers/asset.provider.dart:69]: yeah, that makes sense!

This is sort of non-trivial to fix without ruining upload/backup performance on large collections. I need to think about either making the state a HashSet or aggregating  uploaded assets and updating only in batches after a certain timeout etc.

**Comment by zoodyy** [comment]
> Hi @zoodyy, thank you for another high-quality and high-value PR.

Thanks for reviewing @alextran1502 

>     1. The initial loading speed of the app is fantastic. I think this is from the fewer assets that need to be loaded if the assets exist on the device, correct?

Great! However, this is certainly not the root cause. All assets are still loaded from the server and additionally local assets are added (and the checks to ensure duplicates are removed) - this take compute time and should make it slower to fetch the list of assets. But it does not effect rendering performance at all.
What I did to improve rendering performance is to remove layers of resizing the thumbnails (from 250px to 200px back to 300px for rendering). I guess this explains the faster loading speed.

>     2. I saw we have three different icons to indicate the state of an asset. I think you planned to use two icons, "cloud with the cross mark - indicates the asset available locally and not on remote" and "cloud with the check mark - indicates the asset available remotely." Regarding the other icon - "just a cloud" I don't know what state the asset is in. I observed it in my testing instance, and the asset is confirmed to be available locally and remotely. Please see the screenshot below

I intentionally use three different icons to indicate the state of an asset:

1. "cloud with check mark": Asset was uploaded from this device and is now backed up in the cloud/server
2. "cloud with cross mark": Asset is only available locally and has not yet been backed up
3. "cloud": Asset is only available in the cloud and was uploaded from some other device (like the web client)

I opted to show three states because in the future I plan to connect local assets (with their backup info from the server). The "cloud" icon could also indicate assets that have been backed up but are now deleted on the local device to free up space. Since locally available and remotely available assets can behave slightly different, I think it's nice to differentiate the two states indicating an asset is available remotely ("cloud") and some times both remote and local at the same time "cloud with check mark"


>     3. Now that we need to use the icon to show the state of the photos. I think the icon will need to be presented at all times and not optional anymore - @matthinc, thoughts?

I think we can make it default to show the icons, but it is still nice to disable if a user simply does not care?

File: mobile/lib/modules/album/views/create_album_page.dart [hunk 3]
```diff
@@ -166,7 +166,7 @@ class CreateAlbumPage extends HookConsumerWidget {
                 return GestureDetector(
                   onTap: _onBackgroundTapped,
                   child: SharedAlbumThumbnailImage(
-                    asset: selectedAssets.toList()[index],
+                    asset: selectedAssets.elementAt(index),
```
Comment (alextran1502) [inline mobile/lib/modules/album/views/create_album_page.dart:169]: Such an eleant change!

File: mobile/lib/modules/asset_viewer/views/gallery_viewer.dart [hunk 4]
```diff
@@ -52,15 +51,15 @@ class GalleryViewerPage extends HookConsumerWidget {
...
     getAssetExif() async {
-      assetDetail = await ref
-          .watch(assetServiceProvider)
-          .getAssetById(assetList[indexOfAsset].id);
+      if (assetList[indexOfAsset.value].isRemote) {
+        assetDetail = await ref
+            .watch(assetServiceProvider)
+            .getAssetById(assetList[indexOfAsset.value].id);
+      } else {
+        // TODO local exif parsing?
```
Comment (alextran1502) [inline mobile/lib/modules/asset_viewer/views/gallery_viewer.dart:60]: This would be amazing, though I don't know if it should only happen when the user click the more info button since we will have to open and parse the file through an EXIF parser

File: mobile/lib/modules/backup/background_service/background.service.dart [hunk 5]
```diff
@@ -50,6 +50,11 @@ class BackgroundService {
       _Throttle(_updateProgress, notifyInterval);
   late final _Throttle _throttledDetailNotify =
       _Throttle(_updateDetailProgress, notifyInterval);
+  Completer<bool> _hasAccessCompleter = Completer();
```
Comment (alextran1502) [inline mobile/lib/modules/backup/background_service/background.service.dart:53]: I've never heard of `Completer` before, does it work similar to a `Promise` in Javascript or `Result` type in Rust?

File: mobile/lib/modules/backup/models/hive_backup_albums.model.dart [hunk 6]
```diff
@@ -46,6 +46,15 @@ class HiveBackupAlbums {
     );
   }
 
+  HiveBackupAlbums deepCopy() {
```
Comment (alextran1502) [inline mobile/lib/modules/backup/models/hive_backup_albums.model.dart:49]: What purpose will deepCopy will be served?

File: mobile/lib/modules/home/services/asset.service.dart [hunk 7]
```diff
@@ -1,47 +1,99 @@
...
+import 'package:immich_mobile/constants/hive_box.dart';
+import 'package:immich_mobile/modules/backup/background_service/background.service.dart';
+import 'package:immich_mobile/modules/backup/services/backup.service.dart';
+import 'package:immich_mobile/shared/models/asset.dart';
 import 'package:immich_mobile/shared/providers/api.provider.dart';
 import 'package:immich_mobile/shared/services/api.service.dart';
 import 'package:openapi/api.dart';
+import 'package:photo_manager/src/types/entity.dart';
+
+import '../../backup/models/hive_backup_albums.model.dart';
```
Comment (alextran1502) [inline mobile/lib/modules/home/services/asset.service.dart:15]: Can we change this to the format similar to the import above, I hope to keep things in uniform as we go

File: mobile/lib/modules/search/providers/search_result_page.provider.dart [hunk 8]
```diff
@@ -61,12 +62,11 @@ final searchResultGroupByDateTimeProvider = StateProvider((ref) {
   var assets = ref.watch(searchResultPageProvider).searchResult;
 
   assets.sortByCompare<DateTime>(
-    (e) => DateTime.parse(e.createdAt),
+    (e) => e.createdAt,
```
Comment (alextran1502) [inline mobile/lib/modules/search/providers/search_result_page.provider.dart:65]: I am wondering here what would happen if the parsing failed in the `Asset` model?

**Review by alextran1502 — COMMENTED** [review]
Hi Fynn, I left some additional comments after looking the code in detailed!

Thank you so much for your work!

File: mobile/lib/modules/album/services/album.service.dart [hunk 9]
```diff
@@ -29,7 +30,7 @@ class AlbumService {
 
   Future<AlbumResponseDto?> createAlbum(
     String albumName,
-    Set<AssetResponseDto> assets,
+    Iterable<Asset> assets,
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/modules/album/services/album.service.dart:33]: This allows to call `createAlbum` and `createAlbumWithGeneratedName` with any collection of assets that is iterable (List, Set and most importantly a "stream" of assets produced by e.g. `list.where(...)` or `list.map(...)` without needlessly allocating an intermediate List or Set.
tldr: performance optimization to reduce memory usage and number of memory allocations

File: mobile/lib/modules/asset_viewer/views/gallery_viewer.dart (hunk 4 above)
Comment (zoodyy) [reply mobile/lib/modules/asset_viewer/views/gallery_viewer.dart:60]: There are basically two options:

1. When finding a new local asset on the device, always load the original file and parse the exif and store the info in some local state. -> only makes sense if we actually keep a local state (the current cache is insufficient!) in a database. Still wasteful if the user never looks at the details and the images are uploaded to the server and parsed again.
2. Only load the file and parse exif if the user actually requests that information. -> might take a few milliseconds but should be fast enough to almost be unnoticeable.

We will decide an what to do in a follow-up PR :-)

File: mobile/lib/modules/backup/background_service/background.service.dart (hunk 5 above)
Comment (zoodyy) [reply mobile/lib/modules/backup/background_service/background.service.dart:53]: It's somewhat related to `Promise`. A `Completer` is the low-level building block in dart to create `Future`s that can be `await`ed from callback-based (or similar) code.

File: mobile/lib/modules/backup/models/hive_backup_albums.model.dart (hunk 6 above)
Comment (zoodyy) [reply mobile/lib/modules/backup/models/hive_backup_albums.model.dart:49]: It allows to get a copy that is actually safe to modify without changing the global state of `HiveBackupAlbums`.

Hive does not work like it's API indicates that it works. When you do:

1.  `info = box.get(backupInfoKey);`
2.  modify a few entries in `info`
3. ignore `info` and do nothing with it anymore
4. in some other method load `otherInfo = Hive.box<HiveBackupAlbums>(hiveBackupInfoBox).get(backupInfoKey)`

You would expect `info != otherInfo`. However, the changes made to `info` in 2. are also visible in `otherInfo` although it was not saved/put back into the box!

To guard against this, a deep copy of the arrays in `HiveBackupAlbums` is required.

File: mobile/lib/modules/home/services/asset.service.dart (hunk 7 above)
Comment (zoodyy) [reply mobile/lib/modules/home/services/asset.service.dart:15]: good catch! I need to check if I can configure VSCode to always do package imports when importing something automatically.

File: mobile/lib/modules/home/services/asset.service.dart [hunk 10]
```diff
@@ -1,47 +1,99 @@
...
+
+      assets.addAll(newLocalAssets.map((e) => Asset.local(e)));
     } catch (e) {
       debugPrint("Error [getAllAsset]  ${e.toString()}");
-      return null;
+    }
+    return assets;
+  }
+
+  Future<List<AssetEntity>> _getLocalAssets(bool urgent) async {
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/modules/home/services/asset.service.dart:59]: The idea is the following:

When the user opens the app and the cached state is invalid, no assets are shown until `getAllAsset` completes and updates the state, In case the background service is running, this means that the user will not see any assets for minutes (frustrated, killing the app etc., reporting an issue..).
To guard against this, `urgent` can be set to true when the cached state is invalid. This activates a timeout when reading local assets. If it takes to long for the background service to finish, only remote assets are loaded and shown to the user (until the background service finishes and some action reloads the assets)

File: mobile/lib/modules/home/services/asset.service.dart [hunk 11]
```diff
@@ -1,47 +1,99 @@
...
-  Future<List<AssetResponseDto>?> getAllAsset() async {
+  Future<List<Asset>> getAllAsset({bool urgent = false}) async {
+    final List<Asset> assets = [];
     try {
-      return await _apiService.assetApi.getAllAssets();
+      final Future<List<AssetResponseDto>?> remoteTask =
+          _apiService.assetApi.getAllAssets();
+      final Iterable<AssetEntity> newLocalAssets;
+      final List<AssetEntity> localAssets = await _getLocalAssets(urgent);
+      final List<AssetResponseDto> remoteAssets = await remoteTask ?? [];
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/modules/home/services/asset.service.dart:39]: It allows to fetch remote and local assets concurrently.

If we put an `await` on line `35`, it would do this sequentially (and thus take longer in total)

File: mobile/lib/modules/home/ui/asset_grid/thumbnail_image.dart [hunk 12]
```diff
@@ -87,41 +81,11 @@ class ThumbnailImage extends HookConsumerWidget {
                       )
                     : const Border(),
               ),
-              child: CachedNetworkImage(
-                cacheKey: 'thumbnail-image-${asset.id}',
+              child: ImmichImage(
+                asset,
                 width: 300,
                 height: 300,
-                memCacheHeight: 200,
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/modules/home/ui/asset_grid/thumbnail_image.dart:94]: Agreed. It always feels awesome to delete a bunch on redundant code :D

File: mobile/lib/modules/home/views/home_page.dart [hunk 13]
```diff
@@ -76,9 +78,27 @@ class HomePage extends HookConsumerWidget {
         selectionEnabledHook.value = false;
       }
 
+      Iterable<Asset> remoteOnlySelection() {
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/modules/home/views/home_page.dart:81]: because `return assets.where((a) => a.isRemote);` is neither a Set nor a List. Its an on-the-fly computed "stream" of assets -> avoids memory allocations & reduces memory usage -> performance optimization.

Unless the user selects hundreds or thousands of assets, you probably won't notice. But: it is always faster... unless you iterate the "stream" multiple times afterwards.

File: mobile/lib/modules/search/providers/search_result_page.provider.dart (hunk 8 above)
Comment (zoodyy) [reply mobile/lib/modules/search/providers/search_result_page.provider.dart:65]: The same as before my changes. A `FormatException` is thrown.

File: mobile/lib/shared/providers/asset.provider.dart (hunk 2 above)
Comment (zoodyy) [reply mobile/lib/shared/providers/asset.provider.dart]: fixed: hope I found a good solution
